railties/lib/rails/generators/rails/app/app_generator.rb

```diff
@@ -164,9 +164,12 @@ def config_when_updating
         end
       end
 
-      if !skip_sprockets?
-        insert_into_file "config/application.rb", %(require "sprockets/railtie"), after: /require\(["']rails\/all["']\)\n/
+      unless defined?(Sprockets::Rails)
```

Comment:
Do you think there's a way to avoid adding `sprocket-rails` for applications that have skipped sprockets? Even if it's effectively the same as having the gem transitively required via `rails` in 6.1, we're adding an explicit dependency to the Gemfile that's unnecessary, which doesn't seem great 😕 And users may miss the fact that they need to remove it.

I wonder if we could check `assets_config_exist` as well:
https://github.com/rails/rails/blob/69078b08203e9d3d5421c0f881048a7fe25a5e14/railties/lib/rails/generators/rails/app/app_generator.rb#L120

ie. `unless defined?(Sprockets::Rails) && assets_config_exist`
I suppose apps could have removed the initializer, and then we'd miss adding `sprockets-rails` to the Gemfile, but I imagine that in most cases this would help us determine whether the app was using sprockets in 6.1 or not. What do you think?

Reply:
> Maybe the condition could be checking config/application.rb for either rails/all or sprockets/railtie?

Ooh, that's a great idea! I actually think that solves _not_ adding the gem to apps that had opted to skip sprockets as well? Either:
1) Sprockets are included via `rails/all`, so if we see `rails/all` we know to install `sprockets-rails` + require it in `config/application.rb`
2) The rails frameworks are included individually, and sprockets is also explicitly included via `require "sprockets/railtie"`, in which case we also know to install `sprockets-rails`
3) The rails frameworks are included individually, but we _don't_ see `require "sprockets/railtie"`, in which case we should skip adding the gem _and_ skip inserting the require into `config/application.rb`

Reply:
If a gem is explicitly included in the Gemfile (and does not belong to a group that's not bundle-required), it automatically gets required in the call to `Bundler.require` just below in `config/application.rb`.
Technically it's not the same, as Bundler by default will require "sprockets-rails", which will fail, then try "sprockets/rails", which for Rails frameworks usually does not load the railtie directly, but in that case it does:
https://github.com/rails/sprockets-rails/blob/v3.4.2/lib/sprockets/rails.rb#L2-L3 (this is also the case for the lowest version that Rails has been [requiring for ages](https://github.com/rails/rails/blob/v5.0.0/rails.gemspec#L32), [sprockets-rails v2.0.0](https://github.com/rails/sprockets-rails/blob/v2.0.0/lib/sprockets/rails.rb))

So I think the cleanest thing would be to simply add the gem to the Gemfile, once we've established the app was using it before.

Reply:
Updated to check for `rails/all` or `sprockets/railtie` requires. 

> So I think the cleanest thing would be to simply add the gem to the Gemfile, once we've established the app was using it before.

I agree that it would be cleaner, however I still worry that changing the load order would break things for some people and I lean towards preferring not to break over being slightly cleaner. Since `sprockets-rails` is being added to the bottom of the Gemfile, it would be the last gem required. The `sprockets-rails` [railtie](https://github.com/rails/sprockets-rails/blob/d933dc1df7e1da546256ae65705bf60a28fdb52c/lib/sprockets/railtie.rb) does some things like reopen `Rails::Application::Configuration` that make me nervous about doing so late during boot.

Reply:
>I agree that it would be cleaner, however I still worry that changing the load order would break things for some people and I lean towards preferring not to break over being slightly cleaner.

It's important not to add unnecessary things because they're worrying. We should make sure it's fine to load it later, then do it, or show it's not fine and require manually. Clean and correct is better than less clean and worried about not being correct but we don't know for sure.

But to be honest in that case I don't mind too much adding that line.